Add HTML audio containers and codecs

ddbeck opened the pull request:
Summary: a partial attempt to get audio containers and codec data into BCD, borrowing from @a2sheppy 's [previous work](#3446).

While triaging open PRs, I was trying to imagine how we might merge @a2sheppy's data from #3446. For a number of reasons, I think changing the schema is unlikely to happen soon. In the interim, I'd like to see us migrate that data. With that in mind, I took inspiration from the way we handled events, by naming them in the form `<eventname>_event`. Why couldn't we do likewise for containers and codecs?

This PR takes Sheppy's data (I included him has a coauthor on the commit—he did all the work) and restructures it, using standard `__compat` statements and renaming the features throughout to use a `<containername>_container` and `<codecname>_codec` naming scheme. I've also added human-readable descriptions.

Having generated the JSON, this seems pretty legible to me, but I'd be interested in other perspectives. There's also a couple of issues which I hadn't really attempted to resolve (thus the draft PR):

* ~The codec data does not have status blocks. I'm not sure what statuses would be correct. The HTML spec doesn't demand anything and I'm not sure what other specs would be relevant.~ Never mind; see https://github.com/mdn/browser-compat-data/pull/4193#issuecomment-500441936.
* The PR would introduce a bunch of new `null` data points, which I'd like to avoid. Does anyone know of sources of data for containers and codecs apart for browsers other than Firefox?

a2sheppy commented:
> The PR would introduce a bunch of new `null` data points, which I'd like to avoid. Does anyone know of sources of data for containers and codecs apart for browsers other than Firefox?

We're in the process of getting the other browser organizations to contribute that information.

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
             }
           }
+        },
+        "3gp_container": {
```
I'm thinking we should use something more distinctive for these, such as `media_container_3gp`.

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
+              "deprecated": false
+            }
+          },
+          "aac_codec": {
```
And then `media_codec_aac`...

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
+          },
+          "aac_codec": {
+            "__compat": {
+              "description": "<code>adts</code> codec",
```
ADTS

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
+          },
+          "amr_codec": {
+            "__compat": {
+              "description": "<code>adts</code> codec",
```
AMR, not adts

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
+          },
+          "flac_codec": {
+            "__compat": {
+              "description": "<code>adts</code> codec",
```
FLAC, not adts

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
+          },
+          "mp3_codec": {
+            "__compat": {
+              "description": "<code>adts</code> codec",
```
MP3, not adts

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
+          },
+          "opus_codec": {
+            "__compat": {
+              "description": "<code>adts</code> codec",
```
Opus, not adts

a2sheppy commented on html/elements/audio.json:
```diff
@@ -578,6 +578,3416 @@
+          },
+          "pcm_codec": {
+            "__compat": {
+              "description": "<code>adts</code> codec",
```
And so on... :)

ddbeck commented:
@a2sheppy thanks for this review! I've fixed the descriptions to match their features (not sure how they got mismatched from the original, but it's fixed now) and I've changed the names of the media features to end with `_media_container` and `_media_codec`. I didn't prepend like you suggested for similarity to events (where we append `_event`). But maybe that's an undesirable consistency (perhaps @Elchi3 has some thoughts on it).

I also restored some status data that got lost in the shuffle.

I'd appreciate another look. 😄

bershanskiy commented:
@ddbeck The file is rather large, 3725 lines. I believe it would be much easier to navigate if it was split up into smaller pieces. How about making a tiny file `html/elements/audio/audio.json` with only its compatibility and putting all the data in separate files and folders? For example,
- `html/elements/audio/audio.json` - just tag support
- `html/elements/audio/attributes/` (10 files) - each attribute separately
  - `autoplay.json`
  - `buffered.json`
  - ...
- `html/elements/audio/media_codec/`
- `html/elements/audio/container/`

On the other hand, codec and container support are usually related so it might make sense to put container and corresponding codec into the same file.

This structure should be more intuitive because realistically support for a given attribute (e.g., `muted`) should not be dependent on codec support (e.g., MP3). Also, it should be easier to navigate because each file is less than 100 lines of code.

Elchi3 commented:
This PR needs updating to no longer add `edge_mobile`. Sorry about that!

ddbeck commented:
@Elchi3 Edge removed.

Also, did you have any thoughts on @bershanskiy's proposal? I didn't think this was an especially large file, but maybe `window` isn't a good example to follow. 😆

Elchi3 commented:
Tbh, I have no strong opinion about the splitting of this into smaller files or not. I think what matters most here is how the trees should look like which we are going to export. From reading @bershanskiy, it sounds like different trees are proposed. Right now this PR does:

- html
  - elements
    - audio
      - autoplay
      - other attributes
      - flac_media_container
        - some codec
 

And @bershanskiy's file organization makes it look more like 

- html
  - elements
    - audio
      - attributes
        - autoplay
      - media_codecs
        - some codec
      - media_containers
        - come container

I believe that no other html elements have their attributes under an `attributes` tree, so while it is more explicit, it would be inconsistent with the rest of the HTML compat data.

bershanskiy commented:
@ddbeck @Elchi3 I was suggesting changing the folder structure while maintaining the current tree structure. I just thought may be it would make sense to put all attributes into a separate folder, but now I see that this would probably add more confusion than it's worth.

I don't want to prolong this PR any further and fall into [bike-shedding](https://en.wikipedia.org/wiki/Law_of_triviality) so let's just forget about that proposal.

a2sheppy commented:
I'm reviewing this. It may take some time.

jpmedley commented:
I don't know if it's relevant here, but Chrome doesn't do its own decoding. It compiles in an open source codebase called [ffmpeg](https://ffmpeg.org/), and uses [a build script to specify which codecs](https://cs.chromium.org/chromium/src/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py?l=573) to include. 

Given the number of codecs that exist, I'm concerned about the potential for even separate audio and video files to become unmanageable. My ffmpeg book, written 7 years ago, contains 14 pages of them.

a2sheppy commented:
@jpmedley 	- That's an interesting point, and I don't know if I have a great answer. I definitely feel strongly that offering a list of codecs supported by each browser is worthwhil, especially if it helps to formulate where the overlap is among various browsers for different content type scenarios.

I wonder if it would be worth our while to build a script that can either interpret the build scripts or other data used by each browser when configuring its supported codecs (whether that's done at build time or otherwise). Some mechanism to more automatically generate tables of supported codecs and configurations.

ddbeck commented:
@Elchi3 do you have an opinion on the practicality of including container data?

(Also a note to self: if this is something we intend to add, then I'll need to update the new data guidelines doc in this PR.)

a2sheppy commented:
Of course, the option still remains to pull this information out and keep it in a totally separate file, too, in order to avoid cluttering up the BCD data with it.

Elchi3 commented:
> @Elchi3 do you have an opinion on the practicality of including container data?

Not really, to be honest. I'm not a subject matter expert when it comes to media or codec topics. That said, I'd trust Sheppy and Joe on making the call for MDN and BCD here. If you think MDN readers and/or BCD consumers would benefit from media codec/container data then I'm happy to add it here.

The only constrain that I have is that you should be able to work with the current BCD schema and not invent a new one. Also, if you have found a way to represent media codecs/containers, we should (as Daniel says) document the data guideline you've come up with, so that others can understand and contribute to it.

jpmedley commented:
I've largely stayed out of this because I mainly view it as your call. Given your comment, maybe I should venture in. 

First, if you're going to document everything, the topic is complex enough and deep enough that it deserves its own schema because you'll want a way to programmatically enforce consistency. Containers can support multiple codecs and browser support isn't about the codec. It's about the intersection of a container and a codec. What didn't occur to me until I wrote that last sentence is how that impacts file size. With ordinary browser features, the number of BCD entries is browsers x features. With media it's browsers x containers x codecs. 

The question as to whether it should be done is a judgement call that MDN management will need to make. On the one hand, the goal of having MDN complete and accurate with regard to all web technologies is a worthy one, one that I support. On the other hand, some specialties are rabbit holes with an early point of diminishing returns. 

Video is a case in point. It's a specialty that is itself as deep as the rest of web development, so much so that Google's media people recommend using [Shaka](https://github.com/google/shaka-player/) or [another media player](https://alternativeto.net/software/shaka-player/) rather than trying to build their own. That means that for the average web developer, codec support information is going to come from the player documentation. Much the same can be said of video hosting services such as YouTube or Vimeo.

I would hesitate to make any decisions about containers and codecs without regard to the whole of the media landscape. Off the top of my head, that landscape would look like the list below. I'd sanity check this with video expert.

* Focus on the codecs with the widest support and highest compression ratios. The reason for the former is obvious. The reason for the latter is to minimize use of users' mobile data. (I want to call out AV1. Though it may be new enough to have low browser support, it was designed specifically for a high compression ratio (among other things). I'll be surprised if it isn't widely adopted.)
* Make sure the basics of `<video>` and `<audio>` tags are covered. 
* Provide guidance on when merely embedding a video isn't enough and transitioning to an embedded player is required.
* Provide guidance on when video distribution requires the use of a CDN.
* Provide basic descriptions of and links to common media players.

Elchi3 commented:
Thanks for your insights, Joe. I think @a2sheppy and @chrisdavidmills are the ones to make a call here.

Elchi3 commented:
@chrisdavidmills and @a2sheppy are you planning to comment and follow-up on this? 
If not, we'll close this for now.

chrisdavidmills commented:
@Elchi3 I don't have a huge amount of time to pick this apart right now, and I don't think anyone else on the team does (including sheppy). So my thought would be to merge this as it is for now (it will work for now), try it out on MDN, and then think about we can improve on it in the future.

Elchi3 commented:
It seems this has now been done without putting things into BCD, which is fine with me.
Originally the compat data was on https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats but it now redirects.

https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers contains compat information, but it is in the wiki page rather than in BCD.

Joe said this:
> On the one hand, the goal of having MDN complete and accurate with regard to all web technologies is a worthy one, one that I support. On the other hand, some specialties are rabbit holes with an early point of diminishing returns.

It seems like this is in fact a special rabbit hole and we've decided to not go further down. I'm therefore closing this PR unmerged. This special compat data can be done in the wiki. I've not heard anyone wanting to access it programmatically.
